pkg/controllers/allocation/bind.go

```diff
@@ -27,14 +29,23 @@ import (
 	"k8s.io/client-go/util/workqueue"
 	"knative.dev/pkg/logging"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
 )
 
-type Binder struct {
-	KubeClient   client.Client
-	CoreV1Client corev1.CoreV1Interface
+type Binder interface {
```

Comment:
Curious why you promoted this to an interface? Do we expect multiple implementations?

Reply:
It facilitates using the decorator pattern, with the metric logic being in a separate decorate type.

Reply:
I worry a bit about this pattern for structs that have more methods. You could use struct embedding to achieve a similar mechanism without requiring full method overloading. 
```
type DecoratedBinder {
   Binder // embedded struct, all methods are promoted to DecoratedBinder
}

func (b *Binder) foo() {} 
func (b *Binder) bar() {}

func (d *DecoratedBinder) foo() { 
  // decorate before
  d.Binder.foo()
  // decorate after
}
```
In this case, bar is not overloaded, so anyone calling `decoratedBinder.bar()` will get the undecorated implementation.